common/src/main/java/org/mcaccess/minecraftaccess/features/inventory_controls/GroupGenerator.java

```diff
@@ -358,32 +359,30 @@ public static List<SlotsGroup> generateGroupsFromSlots(AbstractContainerScreenAc
-                        .get(l);
-
-                int m = j + 1;
-                if (optional.isEmpty()) break;
-                StringBuilder clueText = new StringBuilder(Component.translatable("container.enchant.clue", Enchantment.getFullname(optional.get(), l)).withStyle(ChatFormatting.WHITE).getString());
+            for (int i = 0; i < 3; i++) {
+                int enchantmentId = enchantmentScreenHandler.enchantClue[i];
+                int requiredLevel = enchantmentScreenHandler.costs[i];
+                int level = enchantmentScreenHandler.levelClue[i];
+                int cost = i + 1;
+
+                Optional<Holder.Reference<Enchantment>> enchantment = Minecraft.getInstance().level.registryAccess().lookupOrThrow(Registries.ENCHANTMENT).get(enchantmentId);
+                if (enchantment.isEmpty()) break;
+
+                StringBuilder clueText = new StringBuilder(Component.translatable("container.enchant.clue", Enchantment.getFullname(enchantment.get(), level)).getString());
                 if (!PlayerUtils.isCreative()) {
-                    if (Minecraft.getInstance().player.experienceLevel < k) {
-                        clueText.append(Component.translatable("container.enchant.level.requirement", enchantmentScreenHandler.costs[j]).withStyle(ChatFormatting.RED).getString());
+                    if (Minecraft.getInstance().player.experienceLevel < requiredLevel) {
+                        clueText.append(Component.translatable("container.enchant.level.requirement", requiredLevel).getString());
                     } else {
-                        MutableComponent mutableText = m == 1 ? Component.translatable("container.enchant.lapis.one") : Component.translatable("container.enchant.lapis.many", m);
-                        clueText.append(mutableText.withStyle(i >= m ? ChatFormatting.GRAY : ChatFormatting.RED).getString());
-                        MutableComponent mutableText2 = m == 1 ? Component.translatable("container.enchant.level.one") : Component.translatable("container.enchant.level.many", m);
-                        clueText.append(mutableText2.withStyle(ChatFormatting.GRAY).getString());
+                        String lapisCostText = cost == 1 ? Component.translatable("container.enchant.lapis.one").getString() : Component.translatable("container.enchant.lapis.many", cost).getString();
+                        if (enchantmentScreenHandler.getGoldCount() < cost) lapisCostText = "()" + I18n.get("minecraft_access.other.missing") + ") " + lapisCostText;
+                        clueText.append(lapisCostText);
+                        String xpCostText = cost == 1 ? Component.translatable("container.enchant.level.one").getString() : Component.translatable("container.enchant.level.many", cost).getString();
+                        if (PlayerUtils.getExperienceLevel() < cost) xpCostText = "()" + I18n.get("minecraft_access.other.missing") + ") " + xpCostText;
```

Comment:
This line is quite long and complex, it might be better to put the statement within the if statement on its own line. It may also be simpler to use `String.format()` instead of concatenating four things together. The first closing paren also looks like it may be a typo but I could be wrong.

```suggestion
                        if (PlayerUtils.getExperienceLevel() < cost) {
                            xpCostText = String.format("(%s) %s", I18n.get("minecraft_access.other.missing"), xpCostText);
                        }
```

Reply:
> This line is quite long and complex, it might be better to put the statement within the if statement on its own line. It may also be simpler to use `String.format()` instead of concatenating four things together. The first closing paren also looks like it may be a typo but I could be wrong.

Ugh I forgot about String.format(). Thanks